opamp/observiq/agent_config_manager.go

```diff
@@ -0,0 +1,167 @@
+
+	for configName, managedConfig := range a.configMap {
+		// Read in config file
+		cleanPath := filepath.Clean(managedConfig.ConfigPath)
+		configContents, err := os.ReadFile(cleanPath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading config file %s: %w", configName, err)
+		}
+
+		// Add to contentMap
+		contentMap[configName] = &protobufs.AgentConfigFile{
+			Body:        configContents,
+			ContentType: opamp.DetermineContentType(managedConfig.ConfigPath),
+		}
+
+		configNames = append(configNames, configName)
+	}
+
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: contentMap,
+		},
+	}, nil
+}
+
+// ApplyConfigChanges compares the remoteConfig to the existing and applies changes
+func (a *AgentConfigManager) ApplyConfigChanges(remoteConfig *protobufs.AgentRemoteConfig) (effectiveConfig *protobufs.EffectiveConfig, changed bool, err error) {
+	effectiveConfig, err = a.ComposeEffectiveConfig()
+	if err != nil {
+		return nil, false, err
+	}
+
+	currentConfigMap := effectiveConfig.GetConfigMap().GetConfigMap()
+	remoteConfigMap := remoteConfig.GetConfig().GetConfigMap()
+
+	// No remote config Map
+	if remoteConfigMap == nil {
+		return
+	}
+
+	// loop through all remote configs and compare then with existing configs
+	for configName, remoteContents := range remoteConfigMap {
+		// For security check the log file we want is acceptable
+		if _, ok := acceptableConfigs[configName]; !ok {
+			a.logger.Info("Not support config received skipping", zap.String("config", configName))
+			continue
+		}
+
+		currentContents, ok := currentConfigMap[configName]
+		if !ok {
+			a.logger.Info("Untracked config found", zap.String("config", configName))
+			// We don't current track this config file we should add it
+			changed = true
+
+			// Write out the file
+			if err := os.WriteFile(configName, remoteContents.GetBody(), 0600); err != nil {
+				return nil, false, fmt.Errorf("failed to write new config file %s: %w", configName, err)
+			}
+
+			// Track new config
+			a.AddConfig(configName, &opamp.ManagedConfig{
+				ConfigPath: filepath.Join(".", configName),
+				Reload:     opamp.NoopReloadFunc,
+			})
+			continue
+		}
+
+		// Check to see if file contents are the same
+		if bytes.Equal(currentContents.GetBody(), remoteContents.GetBody()) {
+			continue
+		}
+
+		// Update the config file
+		managedConfig := a.configMap[configName]
+		newContents := remoteContents.GetBody()
+
+		a.logger.Info("Applying changes to config file", zap.String("config", configName))
+		configChanged, err := managedConfig.Reload(newContents)
+		if err != nil {
+			return effectiveConfig, false, fmt.Errorf("failed to reload config: %s: %w", configName, err)
+		}
+
+		changed = changed || configChanged
+	}
+
+	// If files have changed recompute effective config
+	if changed {
+		effectiveConfig, err = a.ComposeEffectiveConfig()
+		if err != nil {
+			return nil, true, err
+		}
+	}
+
+	return
```

Comment:
is this some weird go thing where we don't have to return `effectiveConfig, changed, nil`?

Reply:
forgot about named return values. nevermind